File: tests/selenium-smoke/src/test/java/com/fluxion/smoke/pages/LoginPage.java
package com.fluxion.smoke.pages;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;

import java.time.Duration;
import java.util.List;

/**
 * Page Object for the Login page at /login.
 */
public class LoginPage {

    private final WebDriver driver;
    private final WebDriverWait wait;

    // ── Selectors ────────────────────────────────────────────────────
    private static final By EMAIL_INPUT    = By.cssSelector("input[type='email']");
    private static final By PASSWORD_INPUT = By.cssSelector("input[type='password']");
    private static final By SUBMIT_BUTTON  = By.cssSelector("button[type='submit']");
    private static final By ERROR_BANNER   = By.cssSelector(".login-error, [class*='error']");
    private static final By SPLASH_SCREEN  = By.cssSelector("#splash-screen");

    public LoginPage(WebDriver driver, int waitSeconds) {
        this.driver = driver;
        this.wait = new WebDriverWait(driver, Duration.ofSeconds(waitSeconds));
    }

    public LoginPage navigateTo(String baseUrl) {
        driver.navigate().to(baseUrl.replaceAll("/$", "") + "/login");
        waitForPageToLoad();
        return this;
    }

    private void waitForPageToLoad() {
        // Wait for splash screen to disappear if present
        try {
            wait.until(ExpectedConditions.invisibilityOfElementLocated(SPLASH_SCREEN));
        } catch (Exception e) {
            // Splash screen might not be present, continue
        }
        
        // Wait for the email input to be visible and clickable
        wait.until(ExpectedConditions.visibilityOfElementLocated(EMAIL_INPUT));
        wait.until(ExpectedConditions.elementToBeClickable(SUBMIT_BUTTON));
    }

    public LoginPage enterEmail(String email) {
        WebElement el = driver.findElement(EMAIL_INPUT);
        el.clear();
        el.sendKeys(email);
        return this;
    }

    public LoginPage enterPassword(String password) {
        WebElement el = driver.findElement(PASSWORD_INPUT);
        el.clear();
        el.sendKeys(password);
        return this;
    }

    public LoginPage clickSignIn() {
        // Ensure the button is clickable before attempting to click
        wait.until(ExpectedConditions.elementToBeClickable(SUBMIT_BUTTON));
        driver.findElement(SUBMIT_BUTTON).click();
        return this;
    }

    public LoginPage login(String email, String password) {
        enterEmail(email);
        enterPassword(password);
        clickSignIn();
        return this;
    }

    public boolean isOnLoginPage() {
        try {
            wait.until(d -> d.getCurrentUrl().contains("/login"));
            return true;
        } catch (Exception e) {
            return false;
        }
    }

    public boolean isPageLoaded() {
        try {
            wait.until(ExpectedConditions.visibilityOfElementLocated(EMAIL_INPUT));
            wait.until(ExpectedConditions.visibilityOfElementLocated(SUBMIT_BUTTON));
            return true;
        } catch (Exception e) {
            return false;
        }
    }

    public boolean isErrorDisplayed() {
        try {
            wait.until(d -> {
                List<WebElement> els = d.findElements(ERROR_BANNER);
                return els.stream().anyMatch(e -> e.isDisplayed() && !e.getText().isBlank());
            });
            return true;
        } catch (Exception e) {
            return false;
        }
    }

    public String getPageTitle() {
        return driver.getTitle();
    }
}

File: qa/selenium/pages/login_page.py
# qa/selenium/pages/login_page.py
"""
Page Object for the Login page (/login).
"""
from selenium.webdriver.common.by import By
from .base_page import BasePage


class LoginPage(BasePage):
    """Encapsulates interactions with the Fluxion login page."""

    # ── Locators ─────────────────────────────────────────────────────
    EMAIL_INPUT = (By.CSS_SELECTOR, "input[type='email']")
    PASSWORD_INPUT = (By.CSS_SELECTOR, "input[type='password']")
    LOGIN_BUTTON = (By.CSS_SELECTOR, "button[type='submit']")
    REMEMBER_ME = (By.CSS_SELECTOR, "input[type='checkbox']")
    ERROR_MESSAGE = (By.CSS_SELECTOR, ".login-alert-error, .error, [class*='error']")
    FORGOT_PASSWORD_LINK = (By.LINK_TEXT, "Forgot password?")
    SIGN_UP_LINK = (By.PARTIAL_LINK_TEXT, "Sign up")

    def __init__(self, driver, base_url="http://localhost:5173"):
        super().__init__(driver)
        self.base_url = base_url

    def open_page(self):
        self.open(f"{self.base_url}/login")

    def enter_email(self, email):
        self.type_text(self.EMAIL_INPUT, email)

    def enter_password(self, password):
        self.type_text(self.PASSWORD_INPUT, password)

    def check_remember_me(self):
        checkbox = self.find(self.REMEMBER_ME)
        if not checkbox.is_selected():
            checkbox.click()

    def click_login(self):
        self.click(self.LOGIN_BUTTON)

    def login(self, email, password, remember_me=False):
        """Full login flow: enter credentials and submit."""
        self.open_page()
        self.enter_email(email)
        self.enter_password(password)
        if remember_me:
            self.check_remember_me()
        self.click_login()

    def get_error_message(self):
        try:
            return self.get_text(self.ERROR_MESSAGE)
        except Exception:
            return None

    def is_on_login_page(self):
        return "/login" in self.current_url
